src/Storages/StorageS3.cpp

```diff
@@ -320,6 +335,141 @@ class StorageS3BlockOutputStream : public IBlockOutputStream
 };
 
 
+class StorageS3PartitionedBlockOutputStream : public IBlockOutputStream
+{
+public:
+    StorageS3PartitionedBlockOutputStream(
+        const ASTPtr & partition_by,
+        const String & format_,
+        const Block & sample_block_,
+        ContextPtr context_,
+        const CompressionMethod compression_method_,
+        const std::shared_ptr<Aws::S3::S3Client> & client_,
+        const String & bucket_,
+        const String & key_,
+        size_t min_upload_part_size_,
+        size_t max_single_part_upload_size_)
+        : format(format_)
+        , sample_block(sample_block_)
+        , context(context_)
+        , compression_method(compression_method_)
+        , client(client_)
+        , bucket(bucket_)
+        , key(key_)
+        , min_upload_part_size(min_upload_part_size_)
+        , max_single_part_upload_size(max_single_part_upload_size_)
+    {
+        std::vector<ASTPtr> arguments(1, partition_by);
+        ASTPtr partition_by_string = makeASTFunction(FunctionToString::name, std::move(arguments));
+
+        auto syntax_result = TreeRewriter(context).analyze(partition_by_string, sample_block.getNamesAndTypesList());
+        partition_by_expr = ExpressionAnalyzer(partition_by_string, syntax_result, context).getActions(false);
+        partition_by_column_name = partition_by_string->getColumnName();
+    }
+
+    Block getHeader() const override
+    {
+        return sample_block;
+    }
+
+    void write(const Block & block) override
+    {
+        Block current_block_with_partition_by_expr = block;
+        partition_by_expr->execute(current_block_with_partition_by_expr);
+
+        const auto * key_column = checkAndGetColumn<ColumnString>(current_block_with_partition_by_expr.getByName(partition_by_column_name).column.get());
```

Comment:
Maybe we need to sanitize partition key in some way?

For example following code inserting string containing zero bytes

<details>
<summary>Python code snippet</summary>

```
from clickhouse_driver import Client
ch_client = Client('localhost', 9000)

command = """
INSERT INTO TABLE FUNCTION 
s3('http://localhost:9001/foo/test_{_partition_id}.csv', 
    'minioadmin', 'minioadmin', 'CSV', 'id Int32, val String') 
PARTITION BY val
VALUES
"""

ch_client.execute(command, [
    {'id': 1, 'val': '\x00\x00\x00\x00'},
])
```

</details>

Raises

<details>
<summary>ServerException</summary>

```
ServerException: Code: 499.
DB::Exception: The request signature we calculated does not match the signature you provided. Check your key and signing method.. Stack trace:

0. /build/build_docker/../contrib/libcxx/include/exception:133: Poco::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int) @ 0x11fada0c in /home/vdimir/_tmp/clickhouse-23051
1. /build/build_docker/../src/Common/Exception.cpp:59: DB::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int, bool) @ 0x8f2ac9a in /home/vdimir/_tmp/clickhouse-23051
2. /build/build_docker/../src/IO/WriteBufferFromS3.cpp:233: DB::WriteBufferFromS3::makeSinglepartUpload() @ 0xf03d7e8 in /home/vdimir/_tmp/clickhouse-23051
3. /build/build_docker/../src/IO/WriteBufferFromS3.cpp:112: DB::WriteBufferFromS3::finalizeImpl() @ 0xf03ce0b in /home/vdimir/_tmp/clickhouse-23051
4. /build/build_docker/../src/IO/WriteBufferFromS3.cpp:0: DB::WriteBufferFromS3::finalize() @ 0xf03cd61 in /home/vdimir/_tmp/clickhouse-23051
5. /build/build_docker/../src/Storages/StorageS3.cpp:329: DB::StorageS3BlockOutputStream::writeSuffix() @ 0xf944b28 in /home/vdimir/_tmp/clickhouse-23051
6. /build/build_docker/../src/Storages/StorageS3.cpp:0: DB::StorageS3PartitionedBlockOutputStream::writeSuffix() @ 0xf942c3a in /home/vdimir/_tmp/clickhouse-23051
7. /build/build_docker/../contrib/libcxx/include/exception:176: DB::PushingToViewsBlockOutputStream::writeSuffix() @ 0xf357143 in /home/vdimir/_tmp/clickhouse-23051
8. /build/build_docker/../src/Server/TCPHandler.cpp:581: DB::TCPHandler::processInsertQuery(DB::Settings const&) @ 0xfccf69f in /home/vdimir/_tmp/clickhouse-23051
9. /build/build_docker/../contrib/libcxx/include/functional:2236: DB::TCPHandler::runImpl() @ 0xfcca3cd in /home/vdimir/_tmp/clickhouse-23051
10. /build/build_docker/../src/Server/TCPHandler.cpp:1623: DB::TCPHandler::run() @ 0xfcd5c19 in /home/vdimir/_tmp/clickhouse-23051
11. /build/build_docker/../contrib/poco/Net/src/TCPServerConnection.cpp:57: Poco::Net::TCPServerConnection::start() @ 0x11f215a7 in /home/vdimir/_tmp/clickhouse-23051
12. /build/build_docker/../contrib/poco/Net/src/TCPServerDispatcher.cpp:114: Poco::Net::TCPServerDispatcher::run() @ 0x11f21a3c in /home/vdimir/_tmp/clickhouse-23051
13. /build/build_docker/../contrib/poco/Foundation/include/Poco/ScopedLock.h:36: Poco::PooledThread::run() @ 0x120088d9 in /home/vdimir/_tmp/clickhouse-23051
14. /build/build_docker/../contrib/poco/Foundation/include/Poco/SharedPtr.h:156: Poco::ThreadImpl::runnableEntry(void*) @ 0x120065e6 in /home/vdimir/_tmp/clickhouse-23051
15. start_thread @ 0x76db in /lib/x86_64-linux-gnu/libpthread-2.27.so
16. /build/glibc-OTsEL5/glibc-2.27/misc/../sysdeps/unix/sysv/linux/x86_64/clone.S:97: __clone @ 0x12188f in /usr/lib/debug/lib/x86_64-linux-gnu/libc-2.27.so
```
</details>

It's not a big issue, but I'm not sure that something more dangerous can't happen. Any ideas?

Reply:
There is no strong restrictions on key names in S3 :( And no code for validation in SDK. As for `\x00` symbol, it is not technically forbidden, probably this can be even fixed.

https://docs.aws.amazon.com/AmazonS3/latest/userguide/object-keys.html

Reply:
It's stated that `You can use any UTF-8 character in an object key name`, but we can generate invalid UTF-8 string. Anyway, there's `Safe characters`, do you mind to add validation according to this, throwing exception if not valid string found?

Reply:
We already have people who use unsafe characters like `+` (specifically fixed that) and also there may be other non AWS S3 implementations🤷🏼‍♂️🤷🏼‍♂️

Reply:
But my example above return error and message is not clear what happened